service/gcs/runtime/runc/runc.go

```diff
@@ -427,6 +433,37 @@ func (r *runcRuntime) waitOnProcess(pid int) (int, error) {
 
 func (p *process) Wait() (int, error) {
 	exitCode, err := p.c.r.waitOnProcess(p.pid)
+
+	l := logrus.WithField("cid", p.c.id)
+	l.WithField("pid", p.pid).Debug("process wait completed")
+
+	// If the init process for the container has exited, kill everything else in
+	// the container. We don't do this if the container owns its own pid
+	// namespace, because in that case the container init will be the pid
+	// namespace init, so the kernel will kill everything automatically when it
+	// exits. Excluding this condition allows a use case where two containers
+	// can share cgroups (which is how runc determines what to kill), but still
+	// have their processes terminated properly, by not sharing a pid namespace.
```

Comment:
Nit: I feel like this last sentence is kind of confusing the way its written.

Reply:
Thoughts on how it could be improved? What did you find confusing about it?

Reply:
I assume what you're trying to say is that without this condition it's still possible that processes are terminated properly in the case that two containers share a cgroup iff they do not also share a pid namespace. Is that correct?

Reply:
How about:

That way, if two containers share cgroups but not a pid namespace, then when one container terminates, it will not cause the other container to terminate.

And maybe mention that runc is using the cgroup as the determination up above.

Reply:
I've rephrased the comment. @katiewasnothere is it clearer now?

Reply:
I still think it's a bit confusing but it's not super crucial so we can merge if you think this is best. But I think something like "This prevents the issue where if two containers share cgroups but not a pid namespace (they each own their pid namespace), one container's termination would call runc kill and kill both container's processes" makes it more clear what the original issue was. Are you okay with that?

Reply:
How about this?

```
	// If the init process for the container has exited, kill everything else in
	// the container. Runc uses the devices cgroup of the container ot determine
	// what other processes to kill.
	//
	// We don't issue the kill if the container owns its own pid namespace,
	// because in that case the container kernel will kill everything in the pid
	// namespace automatically (as the container init will be the pid namespace
	// init). This prevents a potential issue where two containers share cgroups
	// but have their own pid namespaces. If we didn't handle this case, runc
	// would kill the processes in both containers.
```

Reply:
^ Slight edit

Reply:
Works for me :)